### ISSUE TITLE
How to use memory efficient attention through xformerconfig?

### ISSUE THREAD

**Issue by sh0416** [facebookresearch/xformers#697]
# ❓ Questions and Help

I've just create my own wrapped model for xformer.

```
class GPT2LMHeadModelXFormers(nn.Module):
    def __init__(self, config: GPT2Config, seq_len: int):
        super().__init__()
        # A list of the encoder or decoder blocks which constitute the Transformer.
        xformer_config = [
            {
                "reversible": False,  # Turn on to test the effect of using reversible layers
                "block_type": "encoder",
                "num_layers": config.n_layer,
                "dim_model": config.n_embd,
                "residual_norm_style": "pre",
                "position_encoding_config": {
                    "name": "vocab",
                    "seq_len": seq_len,
                    "vocab_size": config.vocab_size,
                },
                "multi_head_config": {
                    "num_heads": config.n_head,
                    "residual_dropout": config.resid_pdrop,
                    "use_rotary_embeddings": False,
                    "attention": {
                        "name": "scaled_dot_product",
                        "dropout": config.attn_pdrop,
                        "causal": True,
                        "seq_len": seq_len,
                        "num_rules": config.n_head,
                    },
                },
                "feedforward_config": {
                    "name": "FusedMLP",  # Use MLP if Triton is not available
                    "dropout": config.resid_pdrop,
                    "activation": "gelu",
                    "hidden_layer_multiplier": 4,
                },
            }
        ]

        xformer_config = xFormerConfig(xformer_config)
        xformer_config.weight_init = "small"
        print(xformer_config)
        self.encoder = xFormer.from_config(xformer_config)

        # decoder head
        self.ln_f = nn.LayerNorm(config.n_embd)
        self.head = nn.Linear(config.n_embd, config.vocab_size, bias=False)

        self.apply(self._init_weights)
```

I don't know how to apply memory efficient attention through xformerconfig.

Could you help me?

**Comment by sh0416** [comment]
I've just customize the scaled dot product class.

```
@register_attention("memory_efficient_scaled_dot_product", ScaledDotProductConfig)
class MemoryEfficientScaledDotProduct(Attention):
    r"""
    Implementing the Scaled Dot-Product attention proposed in
    `Attention is all you need`_, Vaswani et al.

    .. _`Attention is all you need`: https://arxiv.org/abs/1706.03762v5
    """

    mask: Optional[AttentionMask]

    def __init__(
        self,
        dropout: float = 0.0,
        causal: bool = False,
        seq_len: Optional[int] = None,
        to_seq_len: Optional[int] = None,
        *args,
        **kwargs,
    ):
        super().__init__()

        self.attn_drop = dropout
        self.causal = causal
        self.seq_len = seq_len

        if causal and seq_len is not None:
            self.mask = AttentionMask.make_causal(seq_len, to_seq_len)
        else:
            self.mask = None

        # Properties specific to this attention mechanism
        self.supports_attention_mask = True
        self.supports_key_padding_mask = False

    def forward(
        self,
        q: torch.Tensor,
        k: torch.Tensor,
        v: torch.Tensor,
        att_mask: Optional[Union[AttentionMask, torch.Tensor]] = None,
        *args,
        **kwargs,
    ) -> torch.Tensor:
        r"""
        att_mask    A 2D or 3D mask which ignores attention at certain positions.

                    - If the mask is boolean, a value of True will keep the value,
                        while a value of False will mask the value.

                        Key padding masks (dimension: batch x sequence length) and attention masks
                        (dimension: sequence length x sequence length OR batch x sequence length x sequence length)
                        can be combined and passed in here. Method maybe_merge_masks provided in the utils can be
                        used for that merging.

                    - If the mask has the float type, then an additive mask is expected (masked values are -inf)

        """

        # Convenience, create an attention mask if a tensor was passed
        if att_mask is not None and isinstance(att_mask, torch.Tensor):
            # By default we don't know of the causality, and a check would be expensive
            att_mask = (
                AttentionMask.from_bool(att_mask)
                if att_mask.dtype == torch.bool
                else AttentionMask(att_mask, is_causal=False)
            )

        # Handle a possibly deferred causal mask handling
        mask = self.mask
        if self.causal and self.mask is None:
            mask = AttentionMask.make_causal(
                seq_len=q.shape[-2],
                to_seq_len=q.shape[-2],
                device=q.device,
                dtype=q.dtype,
            )

        # Merge the optional causal mask and the user-provided mask
        if mask is not None:
            mask = mask.to(dtype=q.dtype, device=q.device)

            att_mask = att_mask + mask if att_mask is not None else mask

        # Try to handle a case where the sequence is smaller than the mask
        if (
            att_mask is not None
            and q.shape[-2] == k.shape[-2]
            and q.shape[-2] < att_mask.shape[1]
        ):
            if isinstance(att_mask, AttentionMask):
                att_mask = att_mask.make_crop(seq_len=q.shape[-2])
            else:
                logger.error(
                    "Mismatching sparse attention mask and sequence length."
                    + " Please pad the inputs or adjust the attention mask"
                )
                raise NotImplementedError

        # Attend: (B x nh, S, hs) x (B x nh, hs, S) -> (B x nh, S, S)

        print(att_mask)
        # y = scaled_dot_product_attention(
        #     q=q, k=k, v=v, att_mask=att_mask, dropout=self.attn_drop
        # )
        y = memory_efficient_attention(
            query=q, key=k, value=v, attn_bias=att_mask, p=self.attn_drop
        )
        return y
```

All things are the same except for the last line. However, the attn_bias doesn't get att_mask. Something like xpos.CausalMask seems to be working.. Could you explain what is the difference?

**Comment by danthe3rd** [comment]
For causal you should use:

```python
attn_bias = xformers.ops.LowerTriangularMask()
# ...
y = memory_efficient_attention(
    query=q, key=k, value=v, attn_bias=attn_bias, p=self.attn_drop
)
```

**Comment by sh0416** [comment]
It works although I think I gave up the functionality of att_mask.. Thanks!

**Comment by danthe3rd** [comment]
Merging with a user-provided mask has some limitations (if the size of the sequence is not dividable by 8, some padding might be required + backpropagation won't work I believe?), but in principle it works like this:

```python
attn_bias = xformers.ops.LowerTriangularMaskWithTensorBias(tensor_bias)
# ...
y = memory_efficient_attention(
    query=q, key=k, value=v, attn_bias=attn_bias, p=self.attn_drop
)
```

(see [documentation](https://facebookresearch.github.io/xformers/components/ops.html#xformers.ops.fmha.attn_bias.LowerTriangularMask))

**Comment by sh0416** [comment]
I just curious that memory efficient attention could reduce the GPU memory usage.

Is it?

**Comment by danthe3rd** [comment]
`memory_efficient_attention` should at the same time:
* Reduce memory usage - especially if the sequences (=num tokens) are long (eg > 256). Regular attention uses `O(N^2)` memory (where N is the sequence length), whereas `memory_efficient_attention` uses `O(N)`
* Improve the performance significatively - and even more in the causal setting
